client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java

```diff
@@ -0,0 +1,222 @@
+     *                     the client will not cache any results from the Device Connection service.
+     * @param con The connection to the server.
+     * @param tenantId The tenant to consumer events for.
+     * @param senderCloseHook A handler to invoke if the peer closes the sender link unexpectedly.
+     * @param receiverCloseHook A handler to invoke if the peer closes the receiver link unexpectedly.
+     * @return A future indicating the outcome of the creation attempt.
+     * @throws NullPointerException if any of the parameters other than cache provider is {@code null}.
+     */
+    public static final Future<DeviceConnectionClient> create(
+            final CacheProvider cacheProvider,
+            final HonoConnection con,
+            final String tenantId,
+            final Handler<String> senderCloseHook,
+            final Handler<String> receiverCloseHook) {
+
+        LOG.debug("creating new device connection client for [{}]", tenantId);
+        final DeviceConnectionClientImpl client = new DeviceConnectionClientImpl(con, tenantId);
+        if (cacheProvider != null) {
+            client.setResponseCache(cacheProvider.getCache(DeviceConnectionClientImpl.getTargetAddress(tenantId)));
+        }
+        return client.createLinks(senderCloseHook, receiverCloseHook)
+                .map(ok -> {
+                    LOG.debug("successfully created device connection client for [{}]", tenantId);
+                    return (DeviceConnectionClient) client;
+                }).recover(t -> {
+                    LOG.debug("failed to create device connection client for [{}]", tenantId, t);
+                    return Future.failedFuture(t);
+                });
+    }
+
+    private Map<String, Object> createDeviceIdProperties(final String deviceId) {
+        final Map<String, Object> properties = new HashMap<>();
+        properties.put(MessageHelper.APP_PROPERTY_DEVICE_ID, deviceId);
+        return properties;
+    }
+
+    /**
+     * Invokes the <em>Set Last Known Gateway for Device</em> operation of Hono's
+     * <a href="https://www.eclipse.org/hono/api/device-connection-api">Device Connection API</a>
+     * on the service represented by the <em>sender</em> and <em>receiver</em> links.
+     */
+    @Override
+    public Future<Void> setLastKnownGatewayForDevice(final String deviceId, final String gatewayId, final SpanContext context) {
+        Objects.requireNonNull(deviceId);
+        Objects.requireNonNull(gatewayId);
+
+        final Map<String, Object> properties = createDeviceIdProperties(deviceId);
+        properties.put(MessageHelper.APP_PROPERTY_GATEWAY_ID, gatewayId);
+
+        final Future<DeviceConnectionResult> resultTracker = Future.future();
+        createAndSendRequest(
+                DeviceConnectionConstants.DeviceConnectionAction.SET_LAST_GATEWAY.getSubject(),
+                properties,
+                null,
+                resultTracker,
+                null,
```

Comment:
the intention here is to **never** cache any result, right?

Reply:
yes

Reply:
then IMHO we should also ignore any `Cache-Directive` header when creating the `DeviceConnectionResult` from the response message and should not support supplying a `CacheProvider` when creating the client, don't you think?

Reply:
I've amended the commit accordingly.